src/validation.cpp

```diff
@@ -1148,7 +1209,7 @@ void static InvalidChainFound(CBlockIndex* pindexNew) EXCLUSIVE_LOCKS_REQUIRED(c
 void CChainState::InvalidBlockFound(CBlockIndex *pindex, const CValidationState &state) {
     if (!state.CorruptionPossible()) {
         pindex->nStatus |= BLOCK_FAILED_VALID;
-        m_failed_blocks.insert(pindex);
+        g_blockmetaman.m_failed_blocks.insert(pindex);
```

Comment:
In commit "refactoring: move block metadata into BlockMetadataManager" (155ee896520e51b8a13d6904e3a14b5fa50295c0)

It seems like a step backwards to be replacing a local reference with a global reference. I get that BlockMetadataManager is a singleton shared between multiple chainstates, but I think it's a mistake to add global references to it all over the code. If you could add a `BlockMetadataManager& m_blockmetaman` member to CChainState (also needs forward declaration and initialization in constructor), and replace 
`g_blockmetaman` with `m_blockmetaman` where possible, I think it would help make relationship between the classes clearer, and also be a step in the direction of making validation code more testable and possible to break apart.

Reply:
Good advice, thanks. I've renamed the class `BlockManager` and folded the "global" instance into `ChainstateManager`, which then injects a reference to that instance when constructing chainstates. Each reference lives as `CChainState.m_blockman`.